rust/src/http2/parser.rs

```diff
@@ -0,0 +1,1145 @@
+    while i3.len() > 0 {
+        match http2_parse_headers_block(i3, dyn_headers) {
+            Ok((rem, b)) => {
+                blocks.push(b);
+                debug_validate_bug_on!(i3.len() == rem.len());
+                if i3.len() == rem.len() {
+                    //infinite loop
+                    return Err(Err::Error((input, ErrorKind::Eof)));
+                }
+                i3 = rem;
+            }
+            Err(x) => {
+                return Err(x);
+            }
+        }
+    }
+    return Ok((
+        i3,
+        HTTP2FramePushPromise {
+            padlength,
+            reserved: stream_id.0,
+            stream_id: stream_id.1,
+            blocks,
+        },
+    ));
+}
+
+#[derive(Clone)]
+pub struct HTTP2FrameContinuation {
+    pub blocks: Vec<HTTP2FrameHeaderBlock>,
+}
+
+pub fn http2_parse_frame_continuation<'a>(
+    input: &'a [u8], dyn_headers: &mut Vec<HTTP2FrameHeaderBlock>,
+) -> IResult<&'a [u8], HTTP2FrameContinuation> {
+    let mut i3 = input;
+    let mut blocks = Vec::new();
+    while i3.len() > 0 {
+        match http2_parse_headers_block(i3, dyn_headers) {
+            Ok((rem, b)) => {
+                blocks.push(b);
+                debug_validate_bug_on!(i3.len() == rem.len());
+                if i3.len() == rem.len() {
+                    //infinite loop
+                    return Err(Err::Error((input, ErrorKind::Eof)));
+                }
+                i3 = rem;
+            }
+            Err(x) => {
+                return Err(x);
+            }
+        }
+    }
+    return Ok((i3, HTTP2FrameContinuation { blocks }));
+}
+
+#[repr(u16)]
+#[derive(Clone, Copy, PartialEq, FromPrimitive, Debug)]
+pub enum HTTP2SettingsId {
+    SETTINGSHEADERTABLESIZE = 1,
+    SETTINGSENABLEPUSH = 2,
+    SETTINGSMAXCONCURRENTSTREAMS = 3,
+    SETTINGSINITIALWINDOWSIZE = 4,
+    SETTINGSMAXFRAMESIZE = 5,
+    SETTINGSMAXHEADERLISTSIZE = 6,
+}
+
+impl fmt::Display for HTTP2SettingsId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl std::str::FromStr for HTTP2SettingsId {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let su = s.to_uppercase();
+        let su_slice: &str = &*su;
+        match su_slice {
+            "SETTINGS_HEADER_TABLE_SIZE" => Ok(HTTP2SettingsId::SETTINGSHEADERTABLESIZE),
+            "SETTINGS_ENABLE_PUSH" => Ok(HTTP2SettingsId::SETTINGSENABLEPUSH),
+            "SETTINGS_MAX_CONCURRENT_STREAMS" => Ok(HTTP2SettingsId::SETTINGSMAXCONCURRENTSTREAMS),
+            "SETTINGS_INITIAL_WINDOW_SIZE" => Ok(HTTP2SettingsId::SETTINGSINITIALWINDOWSIZE),
+            "SETTINGS_MAX_FRAME_SIZE" => Ok(HTTP2SettingsId::SETTINGSMAXFRAMESIZE),
+            "SETTINGS_MAX_HEADER_LIST_SIZE" => Ok(HTTP2SettingsId::SETTINGSMAXHEADERLISTSIZE),
+            _ => Err(format!("'{}' is not a valid value for HTTP2SettingsId", s)),
+        }
+    }
+}
+
+//TODOask move elsewhere generic
```

Comment:
Any ideas where I could put this Rust implementation of generic detection versus unsigned integers ?
(and replace detect-engine-uint.h one day)